### DISCUSSION TITLE
Preferred typing of functions

### DISCUSSION THREAD

**Discussion by ReptoxX** [mikro-orm/mikro-orm#6122] [Q&A] [answered]
I'm trying to create a function that accepts an entity, like so:

```ts
function trackUsage(user: User, resource: Resource, amount: number): EntityManager {
    const usage = this.em.create(UserUsage, {
        user: user,
        resource: resource,
        amount: amount,
    }
    this.em.persist(usage);
    return this.em;
}
```

Then i want to use this whenever i need it, but i now have issues whenever i don't fetch the entire User Entity. For example
```ts
const user = this.userRepository.findOne({
    mail: request.mail,
}, { fields: ['id', 'name'] });

await this.trackUsage(user, Resource.apiCalls, 1).flush();
```

How can my trackUsage function accept atleast the Primary Key of the User. Or what type should the "user" be?
It also breaks whenever i populate a relation and just use that user for the function, as it's now `Loaded<User, 'populated'>`

And i don't just want to accept a `number` or `string` of the id, because that gives me no type checking at all.

**Comment by B4nan** [comment]
If you want to accept a partially loaded entity, you need to use the `Loaded` type with the third type argument which describes the partially loaded fields set to `id`. Using `Partial<User>` might work as well (but won't require the id to be present on type level). You no longer have the entity instances on type level with partial loading, since some properties are missing completely from the type.

```ts
interface User {
  readonly id: number;
  readonly name: string;
  readonly parent?: User;
}

function trackUsage(user: Loaded<User, any, 'id'>, amount: number) {
  // ...
}

const u1 = {} as User;
trackUsage(u1, 1);
const u2 = {} as Loaded<User, never, 'id' | 'amount'>;
trackUsage(u2, 1);
const u3 = {} as Loaded<User, 'parent', 'id' | 'amount'>;
trackUsage(u3, 1);
```

> And i don't just want to accept a number or string of the id, because that gives me no type checking at all.

Not sure what you mean by that.

**Reply by ReptoxX** [reply]
Thanks for the quick response!
The solution works when just using it in the where clause, but i can't manually get the `id` of the `user`. So 
```ts
function trackUsage(user: Loaded<User, any, 'id'>) {
    console.log(user.id);
    //               ~~ <- Property 'id' does not exist on type 'Loaded<User, any, "id">
}
```

> Not sure what you mean by that.

I don't want to accept any given string or number to the function to assign the user to. Like the following bad example:

```ts
async function trackUsage(userId: number) {
    const user = await this.userRepository.findOneOrFail({ id: user });
}
```

**Reply by B4nan** [reply, marked as answer]
Ouch, it should be `Loaded<User, never, 'id'>`, then you can access it.

**Reply by ReptoxX** [reply]
I initially believed I had attempted this solution before creating this post and it failed due to populated relations. However, it appears that's no longer the case. 

It works perfectly now!

Thank you for your help, and keep up the excellent work! I really appreciate what you're doing here. :)
